sdks/python/apache_beam/runners/common.py

```diff
@@ -71,99 +72,214 @@ def __init__(self,
+          self.dofn_process = lambda context: fn.process(context, *args)
 
-      class CurriedFn(core.DoFn):
+        class CurriedFn(core.DoFn):
 
-        start_bundle = staticmethod(fn.start_bundle)
-        process = staticmethod(self.dofn_process)
-        finish_bundle = staticmethod(fn.finish_bundle)
+          start_bundle = staticmethod(fn.start_bundle)
+          process = staticmethod(self.dofn_process)
+          finish_bundle = staticmethod(fn.finish_bundle)
 
-      self.dofn = CurriedFn()
+        self.dofn = CurriedFn()
 
-    self.window_fn = windowing.windowfn
-    self.tagged_receivers = tagged_receivers
-    self.step_name = step_name
+      self.window_fn = windowing.windowfn
+      self.tagged_receivers = tagged_receivers
+      self.step_name = step_name
 
-    if state:
-      assert context is None
-      self.context = DoFnContext(self.step_name, state=state)
-    else:
-      assert context is not None
-      self.context = context
+      if state:
+        assert context is None
+        self.context = DoFnContext(self.step_name, state=state)
+      else:
+        assert context is not None
+        self.context = context
 
-    if logging_context:
-      self.logging_context = logging_context
-    else:
-      self.logging_context = get_logging_context(logger, step_name=step_name)
+      if logging_context:
+        self.logging_context = logging_context
+      else:
+        self.logging_context = get_logging_context(logger, step_name=step_name)
 
-    # Optimize for the common case.
-    self.main_receivers = as_receiver(tagged_receivers[None])
+      # Optimize for the common case.
+      self.main_receivers = as_receiver(tagged_receivers[None])
 
   def receive(self, windowed_value):
     self.process(windowed_value)
 
-  def start(self):
-    self.context.set_element(None)
+  def old_dofn_process(self, element):
+    if self.has_windowed_side_inputs and len(element.windows) > 1:
+      for w in element.windows:
+        self.context.set_element(
+            WindowedValue(element.value, element.timestamp, (w,)))
+        self._process_outputs(element, self.dofn_process(self.context))
+    else:
+      self.context.set_element(element)
+      self._process_outputs(element, self.dofn_process(self.context))
+
+  def new_dofn_process(self, element):
+    self.context.set_element(element)
+    arguments, _, _, defaults = self.dofn.get_function_arguments('process')
+    defaults = defaults if defaults else []
+
+    self_in_args = int(isinstance(self.dofn.process, types.MethodType) and \
+        self.dofn.process.im_self is not None and \
+        'self' in arguments)
+
+    # Call for the process function for each window if has windowed side inputs
+    # otherwise we can optimize the runner by calling process for entire window
+    # set in one go.
+    if all(isinstance(s, dict) or s.is_globally_windowed()
+           for s in self.side_inputs):
+      windows = [(window.GlobalWindow(), element.windows)]
```

Comment:
This is based on the existing logic in oldDoFn https://github.com/apache/beam/blob/python-sdk/sdks/python/apache_beam/runners/common.py#L85